### PR TITLE
ci: parallelize graphql build schema, openai build schema, etc

### PR DESCRIPTION
**Pull request by RogerHYang**

### CODE REVIEW

File: .github/workflows/python-CI.yml [hunk 1]
```diff
@@ -122,8 +118,37 @@ jobs:
...
           git diff --exit-code
 
+  build-graphql:
+    name: Build GraphQL Schema
+    needs: changes
+    if: ${{ needs.changes.outputs.phoenix == 'true' }}
+    strategy:
+      matrix:
+        py: [ 3.9, 3.12 ]
+        os: [ ubuntu-latest ]
```
Comment (axiomofjoy) [inline .github/workflows/python-CI.yml:128]: What's stopping us from just building on 3.12?

Comment (axiomofjoy) [reply .github/workflows/python-CI.yml:128]: Not sure it really matters if 3.9 produces a different schema as long as we have one source of truth.

File: .github/workflows/python-CI.yml [hunk 2]
```diff
@@ -105,21 +105,104 @@ jobs:
...
           git diff --exit-code
-      - name: Check notebooks have cleaned output and metadata
+
+  compile-protobuf:
+    name: Compile Protobuf
+    needs: changes
+    if: ${{ needs.changes.outputs.phoenix == 'true' }}
+    strategy:
+      matrix:
+        py: [ 3.11 ]
```
Comment (axiomofjoy) [inline .github/workflows/python-CI.yml:182]: Using 3.11 to make sure it runs in the container?

**Review by axiomofjoy — APPROVED** [review]

File: .github/workflows/python-CI.yml (hunk 2 above)
Comment (RogerHYang) [reply .github/workflows/python-CI.yml:203]: oh i added a comment after your comment. uv is failing for unknown reasons

File: .github/workflows/python-CI.yml [hunk 3]
```diff
@@ -105,23 +110,139 @@ jobs:
...
+          cache-dependency-glob: |
+            pyproject.toml
+            requirements/format-jupyter-notebooks.txt
+            requirements/clean-jupyter-notebooks.txt
+          github-token: ${{ secrets.GITHUB_TOKEN }}
+      - name: Set up `tox` with `tox-uv`
+        run: uv tool install tox --with tox-uv
+#      - name: Format Jupyter notebooks
+#        run: tox run -e format_jupyter_notebooks -- **/*.ipynb
+#      - run: git diff --exit-code
```
Comment (RogerHYang) [inline .github/workflows/python-CI.yml:148]: I can't get this to work for some reason, even though it ran fine locally

**Pull request merged**